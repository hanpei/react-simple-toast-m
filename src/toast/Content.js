import React from 'react';
import PropTypes from 'prop-types';

const Content = ({ children, ...props }) => children({ ...props });
Content.propTypes = { children: PropTypes.func.isRequired };

export default Content;