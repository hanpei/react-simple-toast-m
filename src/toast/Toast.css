.toast {
  position: fixed;
  z-index: 999;
  width: 100px;
  min-height: 100px;
  background: rgba(40, 40, 40, 0.75);
  text-align: center;
  border-radius: 10px;
  color: #ffffff;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 10px;
  top: 50%;
  left: 50%;
  margin-left: -60px;
  margin-top: -60px;
}

.toastContent {
  margin: 0;
  display: block;
  width: 100%;
  text-align: center;
  font-size: 14px;
}

/* mask */
.maskTransparent {
  position: fixed;
  z-index: 990;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}

/* https://github.com/ConnorAtherton/loaders.css/blob/master/loaders.css */

.spinner {
  position: absolute;
  top: 50%;
  left: 50%;
  width: 64px;
  height: 64px;
  margin-left: -32px;
  margin-top: -32px;
  transform: scale(0.6);
}
.spinner div {
  transform-origin: 32px 32px;
  animation: spinner 1.2s linear infinite;
}
.spinner div:after {
  content: ' ';
  display: block;
  position: absolute;
  top: 3px;
  left: 29px;
  width: 5px;
  height: 14px;
  border-radius: 20%;
  background: #fff;
}
.spinner div:nth-child(1) {
  transform: rotate(0deg);
  animation-delay: -1.1s;
}
.spinner div:nth-child(2) {
  transform: rotate(30deg);
  animation-delay: -1s;
}
.spinner div:nth-child(3) {
  transform: rotate(60deg);
  animation-delay: -0.9s;
}
.spinner div:nth-child(4) {
  transform: rotate(90deg);
  animation-delay: -0.8s;
}
.spinner div:nth-child(5) {
  transform: rotate(120deg);
  animation-delay: -0.7s;
}
.spinner div:nth-child(6) {
  transform: rotate(150deg);
  animation-delay: -0.6s;
}
.spinner div:nth-child(7) {
  transform: rotate(180deg);
  animation-delay: -0.5s;
}
.spinner div:nth-child(8) {
  transform: rotate(210deg);
  animation-delay: -0.4s;
}
.spinner div:nth-child(9) {
  transform: rotate(240deg);
  animation-delay: -0.3s;
}
.spinner div:nth-child(10) {
  transform: rotate(270deg);
  animation-delay: -0.2s;
}
.spinner div:nth-child(11) {
  transform: rotate(300deg);
  animation-delay: -0.1s;
}
.spinner div:nth-child(12) {
  transform: rotate(330deg);
  animation-delay: 0s;
}
@keyframes spinner {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

.enter {
  opacity: 0.01;
}
.enterActive {
  opacity: 1;
  transition: opacity 100ms ease-out;
}
.exit {
  opacity: 1;
}
.exitActive {
  opacity: 0.01;
  transition: opacity 100ms ease-out;
}
