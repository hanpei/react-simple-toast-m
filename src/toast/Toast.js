import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { createPortal } from 'react-dom';
import styles from './Toast.css';

function getDisplayName(WrappedComponent) {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component';
}

const createToast = config => WrappedComponent => {
  return class ToastHOC extends Component {
    static displayName = `WithToast(${getDisplayName(WrappedComponent)})`
    state = {
      toastShown: false,
      toastMessage: ''
    };

    constructor() {
      super();
      const div = document.getElementById('react-simple-toast-container');
      if (div) {
        this.modalDiv = div;
      } else {
        this.modalDiv = document.createElement('div');
        this.modalDiv.setAttribute('id', 'react-simple-toast-container');
      }
      document.body.appendChild(this.modalDiv);
    }

    componentWillUnmount() {
      if (this.timerId) {
        clearTimeout(this.timerId);
        this.timerId = undefined;
      }
      if (this.state.toastShown) {
        this.setState({
          toastShown: false
        });
      }
      document.body.removeChild(this.modalDiv);
    }
    showToast = (msg, callback) => {
      this.setState(
        {
          toastShown: true,
          toastMessage: msg
        },
        () => {
          this.timerId = setTimeout(() => {
            this.setState(
              {
                toastShown: false,
                toastMessage: ''
              },
              callback
            );
          }, 1500);
        }
      );
    };
    showLoading = () => {
      this.setState({
        toastShown: true
      });
    };
    hideLoading = () => {
      this.setState({
        toastShown: false
      });
    };

    renderToastComponent(toastMessage) {
      let ToastComponent;
      if (config && config.custom) {
        ToastComponent = config.custom;
      } else {
        ToastComponent = DefaultToast;
      }
      return <ToastComponent message={toastMessage} />;
    }

    render() {
      const { toastShown, toastMessage } = this.state;
      return (
        <Fragment>
          <WrappedComponent
            showToast={this.showToast}
            showLoading={this.showLoading}
            hideLoading={this.hideLoading}
            {...this.props}
          />
          {toastShown &&
            createPortal(
              this.renderToastComponent(toastMessage),
              this.modalDiv
            )}
        </Fragment>
      );
    }
  };
};

const DefaultToast = ({ message }) => {
  return message ? (
    <Fragment>
      <div className={styles.maskTransparent} />
      <div className={styles.toast}>
        <p className={styles.toastContent}>{message}</p>
      </div>
    </Fragment>
  ) : (
    <Fragment>
      <div className={styles.maskTransparent} />
      <div className={styles.toast}>
        <div className={styles.toastContent}>
          <div className={styles.spinner}>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </Fragment>
  );
};

export default createToast;
