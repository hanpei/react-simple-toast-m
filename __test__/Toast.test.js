import React from 'react';
import { shallow } from 'enzyme';
import createToast, { withToast } from '../src/index';
import { DefaultToast } from '../src/toast/Toast';

describe('createToast', () => {
  it('DefaultToast', () => {
    const toast = shallow(<DefaultToast message="hello world" />);
    expect(toast).toMatchSnapshot();
    toast.unmount();
    const toast2 = shallow(<DefaultToast />);
    expect(toast2).toMatchSnapshot();
    toast.unmount();
  });

  it('displayName', () => {
    const App = () => <h1>app</h1>;
    App.displayName = 'nameOfComponent';
    const AppWithToast = createToast()(App);
    const component = shallow(<AppWithToast />);
    expect(AppWithToast.displayName).toBe('WithToast(nameOfComponent)');
  });

  it('injected props', () => {
    jest.useFakeTimers();
    const App = () => <h1>app</h1>;
    const AppWithToast = createToast()(App);
    const component = shallow(<AppWithToast />);
    const app = component.find('App');
    app.props().showLoading();
    expect(component).toMatchSnapshot();
    app.props().hideLoading();
    expect(component).toMatchSnapshot();
    app.props().showToast('abcdefg');
    expect(component).toMatchSnapshot();

    const callback = jest.fn();
    app.props().showToast('abcdefg', callback);
    jest.runAllTimers();
    expect(callback).toBeCalled();
  });

  it('life circle', () => {
    const App = () => <h1>app</h1>;
    const AppWithToast = createToast()(App);
    const fn = jest.spyOn(AppWithToast.prototype, 'componentWillUnmount');
    const component = shallow(<AppWithToast />);
    const app = component.find('App');
    app.props().showToast('abcd');
    app.unmount();
    expect(fn).toBeCalled();
    expect(component).toMatchSnapshot();
  });

  it('custom Toast', () => {
    const App = () => <h1>app</h1>;
    const Custom = () => <h1>custom toast</h1>;
    const AppWithCustomToast = createToast({ custom: Custom })(App);
    const component = shallow(<AppWithCustomToast />);
    expect(component).toMatchSnapshot();
    const app = component.find('App');
    app.props().showToast('abcd');
    expect(app).toMatchSnapshot();
  });

  it('hoc', () => {
    const App = () => <h1>app</h1>;
    const AppWithToast = createToast()(App);
    const component = shallow(<AppWithToast />);
    expect(component).toMatchSnapshot();
    const app = component.find('App');
    const props = component.find('App').props();
    expect(Object.keys(props)).toContain(
      'showToast',
      'showLoading',
      'hideLoading'
    );
    expect(component).toMatchSnapshot();

    const div = document.getElementById('react-simple-toast-container');
    expect(div).toBeDefined();

    component.unmount();
    const div2 = document.getElementById('react-simple-toast-container');
    expect(div2).toBeNull();
  });

  it('withToast', () => {
    const App = () => <h1>app</h1>;
    const Custom = () => <h1>custom toast</h1>;

    const WithToast = withToast();
    const WithCustomToast = withToast({ custom: Custom });
    const WithToastDemo = () => (
      <WithToast>
        {({ showToast }) => (
          <button
            className={styles.btn}
            onClick={() => showToast('hello world again')}
          >
            hi again
          </button>
        )}
      </WithToast>
    );
  
    const component = shallow(<WithToastDemo />);
    expect(component).toMatchSnapshot();

  });
});
