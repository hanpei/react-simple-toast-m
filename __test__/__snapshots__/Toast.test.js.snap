// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createToast DefaultToast 1`] = `
<React.Fragment>
  <div
    className="maskTransparent"
  />
  <div
    className="toast"
  >
    <p
      className="toastContent"
    >
      hello world
    </p>
  </div>
</React.Fragment>
`;

exports[`createToast DefaultToast 2`] = `
<React.Fragment>
  <div
    className="maskTransparent"
  />
  <div
    className="toast"
  >
    <div
      className="toastContent"
    >
      <div
        className="spinner"
      >
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
</React.Fragment>
`;

exports[`createToast custom Toast 1`] = `
<React.Fragment>
  <App
    hideLoading={[Function]}
    showLoading={[Function]}
    showToast={[Function]}
  />
</React.Fragment>
`;

exports[`createToast custom Toast 2`] = `
<App
  hideLoading={[Function]}
  showLoading={[Function]}
  showToast={[Function]}
/>
`;

exports[`createToast hoc 1`] = `
<React.Fragment>
  <App
    hideLoading={[Function]}
    showLoading={[Function]}
    showToast={[Function]}
  />
</React.Fragment>
`;

exports[`createToast hoc 2`] = `
<React.Fragment>
  <App
    hideLoading={[Function]}
    showLoading={[Function]}
    showToast={[Function]}
  />
</React.Fragment>
`;

exports[`createToast injected props 1`] = `
<React.Fragment>
  <App
    hideLoading={[Function]}
    showLoading={[Function]}
    showToast={[Function]}
  />
</React.Fragment>
`;

exports[`createToast injected props 2`] = `
<React.Fragment>
  <App
    hideLoading={[Function]}
    showLoading={[Function]}
    showToast={[Function]}
  />
</React.Fragment>
`;

exports[`createToast injected props 3`] = `
<React.Fragment>
  <App
    hideLoading={[Function]}
    showLoading={[Function]}
    showToast={[Function]}
  />
</React.Fragment>
`;

exports[`createToast life circle 1`] = `
<React.Fragment>
  <App
    hideLoading={[Function]}
    showLoading={[Function]}
    showToast={[Function]}
  />
</React.Fragment>
`;

exports[`createToast withToast 1`] = `<WithToast(Content) />`;
